mountpoint-s3-crt/src/auth/imds_client.rs

```diff
@@ -0,0 +1,127 @@
+    /// client will take ownership.
+    retry_strategy: Option<RetryStrategy>,
+}
+
+impl ImdsClientConfig {
+    /// Create a new [ImdsClientConfig] with default options.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Client bootstrap used for common staples such as event loop group, host resolver, etc.
+    pub fn client_bootstrap(&mut self, client_bootstrap: ClientBootstrap) -> &mut Self {
+        self.inner.bootstrap = client_bootstrap.inner.as_ptr();
+        self.client_bootstrap = Some(client_bootstrap);
+        self
+    }
+
+    /// Retry strategy used to reschedule failed requests
+    pub fn retry_strategy(&mut self, retry_strategy: RetryStrategy) -> &mut Self {
+        self.inner.retry_strategy = retry_strategy.inner.as_ptr();
+        self.retry_strategy = Some(retry_strategy);
+        self
+    }
+}
+
+impl ImdsClient {
+    /// Create a new [ImdsClient]
+    pub fn new(allocator: &Allocator, config: ImdsClientConfig) -> Result<Self, Error> {
+        auth_library_init(allocator);
+
+        // SAFETY: `config` is moved into the [ImdsClient] on success, so `config.inner` (and the values
+        // inside) are guaranteed to live at least as long as this [ImdsClient] does. `allocator` is
+        // guaranteed to be a valid allocator because of the type-safe wrapper.
+        let inner = unsafe { aws_imds_client_new(allocator.inner.as_ptr(), &config.inner).ok_or_last_error()? };
+        Ok(Self { inner })
+    }
+
+    /// Get the ec2 instance type.
+    pub fn get_instance_type<F>(&self, callback: F) -> Result<(), Error>
+    where
+        F: FnOnce(Result<String, Error>) + 'static,
+    {
+        let callback_closure = Box::new(callback);
+        let callback_wrapper = Box::new(ImdsClientGetResourceCallback(Box::new(callback_closure)));
+        let callback_raw_ptr = Box::into_raw(callback_wrapper) as *mut libc::c_void;
+        let get_resource_callback_fn_ptr: Option<unsafe extern "C" fn(*const aws_byte_buf, i32, *mut libc::c_void)> =
+            Some(imds_client_get_resource_callback);
+
+        // SAFETY: `get_resource_callback_fn_ptr` is a guaranteed non-null callback function pointer pass to CRT.
+        // `callback_raw_ptr` is a pointer to a valid rust closure and the binding function will drop it once the
+        // corresponding crt function finish the execution.
+        // See `extern "C" fn imds_client_get_resource_callback`.
+        unsafe {
+            aws_imds_client_get_instance_type(self.inner.as_ptr(), get_resource_callback_fn_ptr, callback_raw_ptr)
+                .ok_or_last_error()
+        }
+    }
+}
+
+impl Drop for ImdsClient {
+    fn drop(&mut self) {
+        // SAFETY: `self.inner` always points to a valid underlying `aws_imds_client`.
+        // when `aws_imdw_client` creation failed, crt frees the allocated memroy
+        // otherwise return the pointer to `aws_imds_client`.
+        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
+    }
+}
```

Comment:
typos. but do we even need the latter sentence? This is one of the more straightforward safety cases I think.

```suggestion
impl Drop for ImdsClient {
    fn drop(&mut self) {
        // SAFETY: `self.inner` always points to a valid underlying `aws_imds_client`.
        unsafe { aws_imds_client_release(self.inner.as_ptr()) };
    }
}
```

Reply:
the latter sentence is actually the reason of  `self.inner` is valid, basically the summarize of function [here](https://github.com/awslabs/aws-c-auth/blob/dd505b55fd46222834f35c6e54165d8cbebbfaaa/source/aws_imds_client.c#L111-L198)

Reply:
Danny's point is that if we're dropping an `ImdsClient`, we know that the `aws_imds_client_new` function you linked to has already succeeded, so the "creation failed" case isn't relevant here.